certstore_linux.go

```diff
@@ -1,14 +1,336 @@
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
+	if certs == nil {
+		C.NSS_Shutdown()
+		return nil, fmt.Errorf("Error %d, closing and returing...\n", int(C.PR_GetError()))
+	}
+	var node *C.CERTCertListNode
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*node)
+		identities = append(identities, &identity)
+	}
+	return identities, nil
+}
+
+type nssIdentity C.CERTCertListNode
+
+func (i *nssIdentity) Signer() (crypto.Signer, error) {
+	return i, nil
+}
+
+func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
+	var der = i.cert.derCert
+	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var der = i.cert.derCert
+	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
+	if certs == nil {
+		return nil, fmt.Errorf("Error %d, cannot list certificates...\n", int(C.PR_GetError()))
+	}
+	var list *C.CERTCertList
+	var node *C.CERTCertListNode
+	list = certs
+	var identities = make([]Identity, 0)
+	var certificates = make([]*x509.Certificate, 0)
+	for node = C.CertListHead(list); C.CertListEnd(node, list) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*node)
+		identities = append(identities, &identity)
+	}
+	certificates = append(certificates, cert)
+	found := true
+	for found != false {
+		found = false
+		cert = certificates[len(certificates)-1]
+		issuer := cert.Issuer.String()
+		for i := 0; i < len(identities); i++ {
+			cert, err = identities[i].Certificate()
+			if cert == nil {
+				return nil, fmt.Errorf("Error %d, cannot fetch certificate...\n", int(C.PR_GetError()))
+			}
+			subject := cert.Subject.String()
+			root := certificates[len(certificates)-1].Subject.String()
+			if subject == issuer && subject != root {
+				certificates = append(certificates, cert)
+				found = true
+				break
+			}
+		}
+	}
+	return certificates, nil
+}
+
+func (i *nssIdentity) Delete() error {
+	C.PK11_DeleteTokenCertAndKey(i.cert, nil)
+	return nil
+}
+
+func (nssIdentity) Close() {
+}
+
+func (i *nssIdentity) Public() (crypto.PublicKey) {
+	var cert, _ = i.Certificate()
+	if cert == nil {
+		return nil
+	}
+	return cert.PublicKey
+}
+
+func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash := opts.HashFunc()
+	if len(digest) != hash.Size() {
+		return nil, errors.New("bad digest for hash")
+	}
+	key := C.PK11_FindKeyByAnyCert(i.cert, nil)
+	if key == nil {
+		return nil, errors.New("cannot find private key")
+	}
+	if key.keyType == C.rsaKey {
+		var pkcs1Prefix = map[crypto.Hash][]byte{
```

Comment:
It feels a bit weird to be hardcoding ASN.1 encoded structures and concatenating it with the digest. I feel like this could use some refactoring.

Reply:
I expect NSS can do all this for us.. just need to find out how

Reply:
I think `Sign` should look something like this, but it's not producing quite the right results for me
```
// Sign implements the crypto.Signer interface.
func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
   hash := opts.HashFunc()
   if len(digest) != hash.Size() {
      return nil, errors.New("bad digest for hash")
   }
   var (
      cert    = C.CERTCertificate(*i)
      key     = C.PK11_FindKeyByAnyCert(&cert, nil)
      secitem C.SECItem
   )
   if key == nil {
      return nil, errors.New("cannot find private key")
   }
   goToNssAlg := map[crypto.Hash]C.SECOidTag{
      crypto.SHA1:   C.SEC_OID_SHA1,
      crypto.SHA256: C.SEC_OID_SHA256,
      crypto.SHA384: C.SEC_OID_SHA384,
      crypto.SHA512: C.SEC_OID_SHA512,
   }
   algID := C.SEC_GetSignatureAlgorithmOidTag(key.keyType, goToNssAlg[opts.HashFunc()])
   if algID == C.SEC_OID_UNKNOWN {
      return nil, errors.New("algorithm unknown")
   }
   rv    := C.SEC_SignData(&secitem, (*C.uchar)(unsafe.Pointer(&digest[0])), C.int(len(digest)), key, algID)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to sign")
   }
   sig := C.GoBytes(unsafe.Pointer(secitem.data), C.int(secitem.len))
   return sig, nil
}
```

Reply:
Or maybe
```
// Sign implements the crypto.Signer interface.
func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
   hash := opts.HashFunc()
   if len(digest) != hash.Size() {
      return nil, errors.New("bad digest for hash")
   }
   var (
      cert    = C.CERTCertificate(*i)
      key     = C.PK11_FindKeyByAnyCert(&cert, nil)
   )
   if key == nil {
      return nil, errors.New("cannot find private key")
   }
   goToNssAlg := map[crypto.Hash]C.SECOidTag{
      crypto.SHA1:   C.SEC_OID_SHA1,
      crypto.SHA256: C.SEC_OID_SHA256,
      crypto.SHA384: C.SEC_OID_SHA384,
      crypto.SHA512: C.SEC_OID_SHA512,
   }
   algID := C.SEC_GetSignatureAlgorithmOidTag(key.keyType, goToNssAlg[opts.HashFunc()])
   if algID == C.SEC_OID_UNKNOWN {
      return nil, errors.New("algorithm unknown")
   }
   ctx := C.SGN_NewContext(algID, key)
   if ctx == nil {
      return nil, errors.New("unable to initialise signing context")
   }
   rv := C.SGN_Begin(ctx)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to initialise signing")
   }
   rv = C.SGN_Update(ctx, (*C.uchar)(unsafe.Pointer(&digest[0])), C.uint(len(digest)))
   if rv != C.SECSuccess {
      return nil, errors.New("unable to sign data")
   }
   var secitem C.SECItem
   rv = C.SGN_End(ctx, &secitem)
   if rv != C.SECSuccess {
      return nil, errors.New("unable to finalise signing")
   }
   sig := C.GoBytes(unsafe.Pointer(secitem.data), C.int(secitem.len))
   return sig, nil
}
``` but that also fails the tests

Reply:
My guess is that many of the "sign" functions are doing the digesting themselves. I'll do some reading to see if I can figure out the correct invocation here.

Reply:
I spent some time playing with this. I think we want to use `SGN_Digest` for signing. Here's the C program I was using for testing this

```C
#include <unistd.h>
#include <stdio.h>
#include <nss.h>
#include <secport.h>
#include <prerror.h>
#include <certt.h>
#include <cert.h>
#include <keyhi.h>
#include <cryptohi.h>
#include <pk11pub.h>

// Create database:
//   mkdir ~/.netscape && certutil -N

// Generate cert:
//   certutil -S -n test1 -x -7 mastahyeti@gmail.com -krsa -tu -s CN=Ben

// Compile:
//   gcc `pkg-config --cflags --libs nss` -o main ./main.c

int main(int argc, char** argv) {
  NSSInitContext *ctxt;
  CERTCertList *list;
  CERTCertificate *cert = NULL;
  SECKEYPrivateKey *key = NULL;
  SECStatus rv;
  SECItem it, digest;

  ctxt = NSS_InitContext("/Users/mastahyeti/.netscape", "", "", SECMOD_DB, NULL, 0);
  if (ctxt == NULL) {
    fprintf(stderr, "NSS_InitContext -> %s\n", PORT_ErrorToString(PORT_GetError()));
    return 1;
  }

  list = PK11_ListCerts(PK11CertListUser, NULL);
  if (list) {
    if (!CERT_LIST_EMPTY(list)) {
      cert = CERT_DupCertificate(CERT_LIST_HEAD(list)->cert);
    }
    CERT_DestroyCertList(list);
  }

  if (cert == NULL) {
    fprintf(stderr, "No cert found\n");
    goto done;
  }

  key = PK11_FindKeyByAnyCert(cert, NULL);
  if (key == NULL) {
    fprintf(stderr, "No private key found\n");
    goto done;
  }

  // Ruby - Digest::SHA256.hexdigest("hello, world!").chars.each_slice(2).map { |q| "0x#{q.join}" }.join(", ")
  unsigned char buf[] = { 0x68, 0xe6, 0x56, 0xb2, 0x51, 0xe6, 0x7e, 0x83, 0x58, 0xbe, 0xf8, 0x48, 0x3a, 0xb0, 0xd5, 0x1c, 0x66, 0x19, 0xf3, 0xe7, 0xa1, 0xa9, 0xf0, 0xe7, 0x58, 0x38, 0xd4, 0x1f, 0xf3, 0x68, 0xf7, 0x28 };
  digest.data = buf;
  digest.len = 32;
  digest.type = siBuffer;

  rv = SGN_Digest(key, SEC_OID_SHA256, &it, &digest);
  if (rv != SECSuccess) {
   fprintf(stderr, "Error signing: %s\n", PORT_ErrorToString(PORT_GetError()));
   goto done;
  }

  write(1, it.data, it.len);

done:
  SECKEY_DestroyPrivateKey(key);
  CERT_DestroyCertificate(cert);

  rv = NSS_ShutdownContext(ctxt);
  if (rv != SECSuccess) {
    fprintf(stderr, "NSS_ShutdownContext -> %d\n", rv);
  }
  return 0;
}
```

Reply:
Something different might need to be done for RSA PSS signatures. I haven't looked at that part yet.

Reply:
Also EC signatures look like they're a bit different too.. 
I don't know if the SGN_* functions will use hardware tokens or not - need to have a play once it's all working

Reply:
Great - I've managed to have some success! `SGN_Digest` is now passing tests. I've kept the code for dealing with EC signatures from the initial PR, but no doubt we can lean on NSS more heavily for that too.

Currently the `Sign()` code is just working - I haven't tidied it up much, but take a nosey anyway if you wish.